lib/charms/lego_base_k8s/v0/lego_client.py

```diff
@@ -0,0 +1,288 @@
+
+    def validate_generic_acme_config(self) -> bool:
+        """Validates generic ACME config."""
+        if not self._email:
+            self.unit.status = BlockedStatus("Email address was not provided")
+            return False
+        if not self._server:
+            self.unit.status = BlockedStatus("ACME server was not provided")
+            return False
+        if not self._email_is_valid(self._email):
+            self.unit.status = BlockedStatus("Invalid email address")
+            return False
+        if not self._server_is_valid(self._server):
+            self.unit.status = BlockedStatus("Invalid ACME server")
+            return False
+        return True
+
+    @abstractmethod
+    def _on_config_changed(self, event: EventBase) -> None:
+        """Validate configuration and sets status accordingly.
+
+        Implementations need to follow the following steps:
+
+        1. Validate their specific configuration, setting the status
+           to `Blocked` if invalid and returning immediately.
+        2. Validate generic configuration by calling
+           `self.validate_generic_acme_config()`, returning immediately
+           if it returns `False`.
+        3. Set the status to `Active` and return.
+
+        Args:
+            event (EventBase): Any Juju event
+
+        Returns:
+            None
+        """
+
+    @staticmethod
+    def _get_subject_from_csr(certificate_signing_request: str) -> str:
+        """Returns subject from a provided CSR."""
+        csr = x509.load_pem_x509_csr(certificate_signing_request.encode())
+        subject_value = csr.subject.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value
+        if isinstance(subject_value, bytes):
+            return subject_value.decode()
+        else:
+            return subject_value
+
+    def _push_csr_to_workload(self, csr: str) -> None:
+        """Pushes CSR to workload container."""
+        self._container.push(path=self._csr_path, make_dirs=True, source=csr.encode())
+
+    def _execute_lego_cmd(self) -> bool:
+        """Executes lego command in workload container."""
+        process = self._container.exec(
+            self._cmd, timeout=300, working_dir="/tmp", environment=self._plugin_config
+        )
+        try:
+            stdout, error = process.wait_output()
+            logger.info(f"Return message: {stdout}, {error}")
+        except ExecError as e:
+            logger.error("Exited with code %d. Stderr:", e.exit_code)
+            for line in e.stderr.splitlines():  # type: ignore
+                logger.error("    %s", line)
+            return False
+        return True
+
+    def _pull_certificates_from_workload(self, csr_subject: str) -> List[Union[bytes, str]]:
+        """Pulls certificates from workload container."""
+        chain_pem = self._container.pull(path=f"{self._certs_path}{csr_subject}.crt")
+        return [cert for cert in chain_pem.read().split("\n\n")]  # type: ignore[arg-type]
```

Comment:
Nitpick: I am curious why this method returns a list and the caller has to use the index to get the certificate, ca, and chain. It would be better to directly make the return data ready.

Another question is: is the order guaranteed to be the same everytime?

Reply:
Again, this charm here is not a owner of this lib, we may want to change this but that does not feel like a charm review criteria